EA-2581: Include correct headers in responses

ikattey opened the pull request.

ikattey committed:
EA-2581: Include correct headers in responses

ikattey committed:
EA-2581: Populate Allow header with matching request methods

ikattey committed:
EA-2581: Accept both application/json and application/ld+json requests

ikattey committed:
Merge branch 'develop' into EA-2581_response_headers

ikattey committed:
Merge remote-tracking branch 'origin/develop' into EA-2581_response_headers

gsergiu commented on entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EMControllerIT.java:
```diff
@@ -392,6 +391,9 @@ void retrieveConceptExternalShouldBeSuccessful() throws Exception {
                 .andExpect(jsonPath("$.id", is(concept.getEntityId())))
                 .andExpect(jsonPath("$.type", is(EntityTypes.Concept.name())));
 
+        checkAllowHeaderForGET(resultActions);
```
can we switch to checkAllowHeaderForDPGP?

ikattey replied:
@gsergiu `/{type}/base/{identifier}.jsonld` is only implemented for GET requests. That's the route pattern used by this test.

gsergiu replied:
You are right here, the issues is in the specs. Indeed for the .jsonld and .xml only the GET method exists .... We will need to update the specs

gsergiu commented on entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EMControllerIT.java:
```diff
@@ -417,6 +419,8 @@ void retrieveConceptWithLanguageFilteringShouldBeSuccessful() throws Exception {
         resultActions.andExpect(jsonPath("$.prefLabel[*]", hasSize(2)))
 		.andExpect(jsonPath("$.altLabel[*]", hasSize(1)));
+
+        checkAllowHeaderForGET(resultActions);
```
can we switch to checkAllowHeaderForDPGP?

ikattey replied:
See my response above.

gsergiu commented on entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EMControllerIT.java:
```diff
@@ -459,12 +461,10 @@ public void retrieveWithContentNegotiationShouldBeSuccessful() throws Exception
         resultActions.andExpect(status().isOk())
-        	.andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+        .andExpect(header().string(HEADER_CONTENT_TYPE,
+                is(CONTENT_TYPE_JSONLD_UTF8)))
```
can we add the checkAllowHeaderForDPGP?

ikattey replied:
This is a valid point. I'll check for the Allow header here.

gsergiu commented on entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EMControllerIT.java:
```diff
@@ -487,6 +487,8 @@ public void retrieveAgentExternalShouldBeSuccessful() throws Exception {
                 .andExpect(jsonPath("$.id", is(agent.getEntityId())))
                 .andExpect(jsonPath("$.type", is(EntityTypes.Agent.name())));
 
+        checkAllowHeaderForGET(resultActions);
```
can we switch to checkAllowHeaderForDPGP?

gsergiu commented on entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EMControllerIT.java:
```diff
@@ -736,6 +741,46 @@ public void changeProvenanceShouldBeSuccessful() throws Exception {
+                is(CONTENT_TYPE_JSONLD_UTF8)))
+                .andExpect(header().exists(HttpHeaders.ETAG))
+                .andExpect(header().string(HttpHeaders.LINK, is(VALUE_LDP_RESOURCE)))
+                .andExpect(header().stringValues(HttpHeaders.VARY, hasItems(containsString(HttpHeaders.ACCEPT))));
```
the delete method must not include a Vary header

gsergiu commented on entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EMControllerIT.java:
```diff
@@ -779,6 +824,9 @@ private String getRetrieveEntityXmlResponse(String requestPath) throws Exception
         		.andExpect(status().isOk())
+                // content-type matches Accept header
+                .andExpect(header().string(HEADER_CONTENT_TYPE,
+                        is(MediaType.APPLICATION_XML_VALUE)))
```
the response should include application/rdf+xml in content type response header

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/config/WebMvcConfig.java:
```diff
@@ -13,17 +13,17 @@
         registry.addMapping("/**")
                 .allowedOrigins("*")
-                .allowedMethods(HttpMethod.GET.name(), HttpMethod.HEAD.name(), HttpMethod.POST.name(), HttpMethod.PUT.name(), HttpMethod.DELETE.name(), HttpMethod.OPTIONS.name())
+                .allowedMethods("*")
```
we can either create the mappings for individual patterns (i.e. /entity/, /entity/*/*,  /entity/*/*/* ) or use the  @CrossOrigin annotation on individual methods

ikattey replied:
Not sure what you mean. This is how CORS is configured – the wildcard is just a shortcut for all HTTP methods.

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/config/WebMvcConfig.java:
```diff
@@ -13,17 +13,17 @@
+                .allowedMethods("*")
                 .allowedHeaders("*")
                 .allowCredentials(false)
+                .exposedHeaders(HttpHeaders.ALLOW)
```
DELETE /entity/{type}/{identifier}/management does not return allow header, but it might be acceptable to include it in all responses

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java:
```diff
@@ -235,58 +245,38 @@ public EMController(EntityRecordService entityRecordService,
-	return createResponse(profile, type, languages, identifier, FormatTypes.xml, HttpHeaders.CONTENT_TYPE_APPLICATION_RDF_XML);
+			// if a valid Accept header value is provided, use it; otherwise return application/rdf+xml
+		    // TODO: handle browser requests with multiple Accept header values
+			String responseContentType = GET_XML_ACCEPT_HEADERS.contains(acceptHeader)
```
I think that we need to return always application/rdf+xml in content type header, even when the request comes with application/xml

gsergiu reviewed (changes requested):
There are still some configurations and header values to be fixed

ikattey committed:
Merge branch 'develop' into EA-2581_response_headers

ikattey committed:
EA-2581: Implement review remarks

- always return application/rdf+xml content-type for XML responses
- do not return any content-type for empty DELETE responses

Commit:
Merge branch 'develop' into EA-2581_response_headers

Conflicts:
	entity-management-web/src/main/java/eu/europeana/entitymanagement/web/BaseRest.java

Commit:
fixed merge conflict

Commit:
#EA-2581 comment out duplicated exception handler
